### PR TITLE
[SPARK-20604][ML] Allow imputer to handle numeric types

### PR DESCRIPTION
**Pull request by actuaryzhang**
## What changes were proposed in this pull request?

Imputer currently requires input column to be Double or Float, but the logic should work on any numeric data types. Many practical problems have integer  data types, and it could get very tedious to manually cast them into Double before calling imputer. This transformer could be extended to handle all numeric types.  

## How was this patch tested?
new test

### CODE REVIEW

**Comment by actuaryzhang** [comment]
@yanboliang @srowen @MLnick @jkbradley 

The example below shows failure of Imputer on integer data. 
```
    val df = spark.createDataFrame( Seq(
      (0, 1.0, 1.0, 1.0),
      (1, 11.0, 11.0, 11.0),
      (2, 1.5, 1.5, 1.5),
      (3, Double.NaN, 4.5, 1.5)
    )).toDF("id", "value1", "expected_mean_value1", "expected_median_value1")
    val imputer = new Imputer()
      .setInputCols(Array("value1"))
      .setOutputCols(Array("out1"))
    imputer.fit(df.withColumn("value1", col("value1").cast(IntegerType)))

java.lang.IllegalArgumentException: requirement failed: Column value1 must be of type equal to one of the following types: [DoubleType, FloatType] but was actually of type IntegerType.

```

**Comment by sethah** [comment]
So the other PR https://github.com/apache/spark/pull/11601 is really long. For reference, I am picking out the relevant discussions to this PR (also someone tell me if there's a better way to link to pr comments :)

@MLnick "what do you think about handling different numeric types in input/output columns? If the input is say IntType, then strategy mode andmedian is ok but mean is somewhat problematic - or are we ok with rounding to and Int? The alternative is the Imputer always appends a Double output column.

I propose we either (a) do the cast back to input type, but if the user selected "mean" and the input type is not Float or Double, log a warning; or (b) only support Float and Double type for this initial version of the Imputer."

@jkbradley "Just catching up now... I like the idea of maintaining the input type. I'm imagining using an Imputer to fill in continuous features with the mean and categoricals with the mode. Later on, we could even check to see if a column is categorical (in the metadata) and throw an exception for mean.

I'd prefer your option (b) to be safe."

@sethah "For reference, I checked scikit-learn and the Imputer class returns floats regardless of inputs. I also checked R package "mlr" and it appears to do the same. One concern with a.) would be if the true median was something like 5.0, but approxQuantile returned 4.999999999. Then, we cast back to IntegerType and return 4. I wasn't able to produce this situation when I briefly experimented with it, and also the median is already approximate, so I'm not sure if this is really a problem."

**Comment by actuaryzhang** [comment]
@sethah Thanks for summarizing the previous discussions. 
What are you suggesting for this PR? I think it makes sense to log a warning when imputing integer types with mean. In addition, perhaps we can set "median" as the default strategy.

**Comment by hhbyyh** [comment]
I imagine most Int features will need to be converted to Double for a Vector, thus returns Double regardless the input type makes sense, which also makes the implementation more straight forward.

**Comment by actuaryzhang** [comment]
@hhbyyh Thanks for the suggestion. I have made a new commit that always casts the input to double and outputs the imputed column as double.

**Comment by actuaryzhang** [comment]
@hhbyyh @sethah @MLnick 
Could you take a look at the new commit? Thanks.

**Comment by actuaryzhang** [comment]
Ping folks for comments/review. Many thanks. 
@viirya @MLnick @jkbradley @hhbyyh @yanboliang @BenFradet

**Comment by MLnick** [comment]
Originally the idea behind only supporting double was as @sethah posted above - there could be some issues with handling of int casting etc. As mentioned originally, we did consider "always cast to double". The only issue with it is the potential for surprising users who may expect the type of the input column to be maintained in the imputation.

Having said that I would be broadly ok with just appending a double output column, provided we update the docstrings / guide to make things very clear.

**Comment by actuaryzhang** [comment]
@MLnick Thanks much for your comments. Yes, I think always returning Double is consistent with Python and R and also other transformers in ML. Plus, as @hhbyyh mentioned, this makes the implementation easier. Would you mind taking a look at the code and let me know if there is any suggestion for improvement? The doc is already updated to make it clear that it always returns Double regardless of the input type.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala [hunk 1]
```diff
@@ -94,12 +94,13 @@ private[feature] trait ImputerParams extends Params with HasInputCols {
...
  * Imputation estimator for completing missing values, either using the mean or the median
  * of the columns in which the missing values are located. The input columns should be of
- * DoubleType or FloatType. Currently Imputer does not support categorical features
+ * numeric type. Currently Imputer does not support categorical features
  * (SPARK-15041) and possibly creates incorrect values for a categorical feature.
  *
  * Note that the mean/median value is computed after filtering out missing values.
  * All Null values in the input columns are treated as missing, and so are also imputed. For
  * computing median, DataFrameStatFunctions.approxQuantile is used with a relative error of 0.001.
+ * The output column is always of Double type regardless of the input column type.
```
Comment (actuaryzhang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala]: @MLnick Here is the note on always returning Double type.

**Comment by actuaryzhang** [comment]
Any committer has a chance to take another look at this PR? Thanks.

**Comment by hhbyyh** [comment]
Shall we pay extra attention to the Int case? E.g. input column contains 
Double.Nan, 
1, 
2. 

The current implementation will return surrogate as 1.5. I'm not sure if it's the expectation for some users. 

It's fine by me but just bring up the issue in case it's missed.

**Comment by actuaryzhang** [comment]
We can log a warning or issue an error if the input column is int and the imputation is by mean.
Would like to know if that's OK with you? @hhbyyh @MLnick

**Comment by felixcheung** [comment]
what's next on this one?

**Comment by actuaryzhang** [comment]
Thanks for following up on this, Felix. 
Still waiting for an agreement on this...
Would like to have more direction on this.

**Comment by holdenk** [comment]
It sounds like folks are generally OK with the approach being taken here if you wanted to update the PR?

**Comment by srowen** [comment]
Reviewing this old one .. I'd favor not changing the column type. When outputting the mean, it should round back to integer types if needed.

**Comment by actuaryzhang** [comment]
Sorry, have not been active here due to other stuff...  
Do folks still feel we should proceed with this PR? 
If so, please suggest any remaining changes and I'll update it. 
@srowen @holdenk @felixcheung @hhbyyh

**Comment by srowen** [comment]
I personally would not favor any change that changes the type of the column. Either let's not do this or change the logic to round the results back to integer types in the case of integer columns.

**Comment by actuaryzhang** [comment]
@srowen If keep the column type intact (return integer for integer input), will you be supporting this PR then?

**Comment by srowen** [comment]
I think that would be pretty reasonable and I'd review it.

**Comment by actuaryzhang** [comment]
@srowen @holdenk 
Made an update to preserve input data type. 
Logic is to convert input to Double internally, compute mean/median, do imputation, and then cast back to original data type. Updated the doc and test accordingly. 
Please take a look. Thanks.

File: mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala [hunk 2]
```diff
@@ -176,6 +178,51 @@ class ImputerSuite extends MLTest with DefaultReadWriteTest {
     assert(newInstance.surrogateDF.collect() === instance.surrogateDF.collect())
   }
 
+  test("Imputer for Numeric with default missing Value NaN") {
+    val df = spark.createDataFrame(Seq(
+      (0, 1.0, 1.0, 1.0),
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala]: I think you don't need the id column?

File: mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala [hunk 3]
```diff
@@ -176,6 +178,51 @@ class ImputerSuite extends MLTest with DefaultReadWriteTest {
...
+      (1, 11.0, 11.0, 11.0),
+      (2, 3.6, 3.6, 3.6),
+      (3, Double.NaN, 5.2, 3.6)
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala]: It's not necessarily true that the mean of the values after casting is equal to the mean after casting, but here it happens to be. That's OK I think. What about checking the case where a non-NaN value is set as the missing value?

File: mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala [hunk 4]
```diff
@@ -176,6 +178,51 @@ class ImputerSuite extends MLTest with DefaultReadWriteTest {
...
+    )).toDF("id", "value1", "expected_mean_value1_double", "expected_median_value1_double")
+
+    val imputer = new Imputer()
+      .setInputCols(Array("value1"))
+      .setOutputCols(Array("out1"))
+
+    val types = Seq(ShortType, IntegerType, LongType, FloatType, DoubleType,
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala]: I think this is fine, but wouldn't necessarily bother with anything but Integer Long, maybe Float. (Not sure how long the test takes to run.)

File: mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala [hunk 5]
```diff
@@ -84,9 +84,15 @@ private[feature] trait ImputerParams extends Params with HasInputCols with HasOu
  * :: Experimental ::
  * Imputation estimator for completing missing values, either using the mean or the median
  * of the columns in which the missing values are located. The input columns should be of
- * DoubleType or FloatType. Currently Imputer does not support categorical features
+ * numeric type. Currently Imputer does not support categorical features
  * (SPARK-15041) and possibly creates incorrect values for a categorical feature.
  *
+ * Note that the input columns are converted to Double data type internally to compute
+ * the mean/median value and impute the missing values, which are then casted back to
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala]: I wouldn't put all of this implementation detail into the docs. I would however note that in the case of integer types and mean imputation, the mean will be cast (truncated) to an integer type. That is, your example is a good one.

**Review by srowen — COMMENTED** [review]
Nice, that's tidy.

File: mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala (hunk 2 above)
Comment (actuaryzhang) [reply mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala]: Removed now. I was just copying examples from existing tests.

File: mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala (hunk 3 above)
Comment (actuaryzhang) [reply mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala]: Added another test to check non-NaN missing value.

File: mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala (hunk 4 above)
Comment (actuaryzhang) [reply mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala]: Thanks. Just keeping IntegerType and LongType now.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala (hunk 5 above)
Comment (actuaryzhang) [reply mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala]: Great suggestion. Streamlined the doc

**Comment by actuaryzhang** [comment]
@srowen Thanks for the review. Made an update that addressed your comments.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**